tensorflow_io/core/kernels/dataset_ops.h

```diff
@@ -152,26 +152,89 @@ class DataInput {
  public:
   DataInput() {}
   virtual ~DataInput() {}
-  virtual Status FromStream(io::InputStreamInterface& s) = 0;
-  virtual Status ReadRecord(io::InputStreamInterface& s, IteratorContext* ctx, std::unique_ptr<T>& state, int64 record_to_read, int64* record_read, std::vector<Tensor>* out_tensors) const = 0;
+  virtual Status ReadReferenceRecord(void* s, IteratorContext* ctx, std::unique_ptr<T>& state, int64 record_to_read, int64* record_read, std::vector<Tensor>* out_tensors) const = 0;
```

Comment:
Would you mind explaining why using `void *` here and the change to using pointers instead of references?

Reply:
@BryanCutler Previously, references works well as we are mostly dealing with file streams (`io::InputStreamInterface& s`). However, for gRPC it is not a file stream and is not returning raw bytes (no `ReadNBytes`). So `InputStreamInterface` is irrelevant here. Changed to pointers so that `nullptr` could be passed for gRPC.

The main purpose is to try to reuse the logic of batch where we need to piece together chunks of tensor.

Reply:
Ok, thanks for the explanation. Having `void*` arguments on public apis make a little nervous though. Do you think putting this as a `protected` member makes sense?

Reply:
@BryanCutler Thanks for the review. The PR has been updated. Now FileInput (and subclass delcared) only need to implement:
```
ReadRecord(io::InputStreamInterface* s,  IteratorContext* ctx, ...)
```

and StreamInput only need to implement:
```
Status ReadRecord(IteratorContext* ctx, ...
```

So `void *` is hidden and will not be touched by any ops.

Please take a look.

Reply:
Sounds good thanks!